File: 735.cpp
/*
模拟栈：
如果当前的没有爆炸加入 stack 中每次拿到 stack 的 top（这里是back）查看是否跟当前的 asteroid 会导致爆炸，没有加入其中，有的话：如果相同 popup（pop_back)，如果绝对值栈顶的更小则 continue跟下一个比，处理没有爆炸的：没爆炸的加入到结果 stack 中
*/
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        vector<int> stack;
        for(int asteroid : asteroids) {
            bool exploded = false;
            while(!stack.empty() && asteroid < 0 && stack.back() > 0){
                if(stack.back() < - asteroid) {
                    stack.pop_back();
                    continue;
                }else if(stack.back() = -asteroid) {
                    stack.pop_back();
                }
                exploded = true;
                break;
            }
            if(!exploded) {
                stack.push_back(asteroid);
            }
        }
        return stack;
    }
};

File: Medium/735.py
# asteroid collision
from typing import List
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []
        for asteroid in asteroids:
            # 向右移动
            if asteroid > 0 or not stack:
                stack.append(asteroid)
                continue
            # 向左移动
            while True:
                if not stack:
                    stack.append(asteroid)
                    break
                if stack[-1] < 0:
                    stack.append(asteroid)
                    break
                # 栈顶向右
                if stack[-1] < abs(asteroid):
                    stack.pop()
                    continue
                if stack[-1] == abs(asteroid):
                    stack.pop()
                    break
                # 栈顶的更大，当前的 asteroid 爆炸
                break
        return stack
    
    def asteroidCollision1(self, asteroids: List[int]) -> List[int]:
        stack = []
        for asteroid in asteroids:
            #针对每一个asteroid去标记的
            will_survive = True
            while will_survive and stack and stack[-1] > 0 and asteroid < 0:
                if stack[-1] > -asteroid:
                    will_survive = False
                elif stack[-1] == -asteroid:
                    will_survive = False
                    stack.pop()
                else:
                    stack.pop()
            if will_survive:
                stack.append(asteroid)
        return stack

